SP-8416: implement timestamps for application retransmissions

przemyslawfierek-silvair opened the pull request.

przemyslawfierek-silvair committed:
amqp: remove identity field

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -222,7 +224,35 @@ static void set_defaults(struct mesh_node *node)
+struct postponed_msg {
+	struct fd_msg *msg;
+	size_t msg_len;
+};
```
```c
struct postponed_msg {
    size_t msg_len;
    struct fd_msg fd_msg;
}

static struct postponed_msg *postponed_msg_new(struct fd_msg *fd_msg, size_t msg_len)
{
    struct postponed_msg *msg = l_malloc(sizeof(*msg) + msg_len);
    msg->msg_len = msg_len;
    memcpy(&msg->fd_msg, fd_msg, msg_len);
    return msg;
}
```
avoids the 2nd `l_malloc` and you don't need `postponed_msg_free` anymore

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
 
 	if (!mesh_model_send(node, src, msg->dst_addr, app_idx, msg->net_idx,
-		      msg->ttl, FD_MSG_IS_SEGMENTED(msg), data_len, msg->data))
+		msg->ttl, FD_MSG_IS_SEGMENTED(msg), data_len, msg->data))
```
I think this should be indented (but with a tab). Separate commit?

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
+	if (!postponed_msg)
+		return;
+
+	if (postponed_msg->msg->timestamp <= current_timestamp) {
```
This should be a `while`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
+		/* to avoid recursive function call, schedule another timeout */
+		l_timeout_modify_ms(timeout, 1);
+	}
+}
```
If you process messages in a loop, you don't need to worry about recurstion:

```c
static void try_send_rc_msg(struct l_timeout *timeout, void *user_data)
{
    struct mesh_node *node = user_data;
    uint64_t current_timestamp = get_timestamp_ms();
    struct postponed_msg *postponed_msg;

    while ((postponed_msg = l_queue_peek_head(node->rc_priority_queue))) {
        if (postponed_msg->msg->timestamp > current_timestamp)
            break;

        send_rc_msg(postponed_msg->msg, postponed_msg->msg_len, node);
        l_queue_pop_head(node->rc_priority_queue);
        l_free(postponed_msg);
    }

    if (postponed_msg) {
        uint64_t diff = postponed_msg->msg->timestamp - current_timestamp;
        l_timeout_modify_ms(timeout, diff);
    }
}
```

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
+	return msg2->msg->timestamp >= msg1->msg->timestamp;
+}
+
+static void try_send_rc_msg(struct l_timeout *timeout, void *user_data)
```
I'd include "flush" somewhere in the function name ;)

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -252,10 +282,70 @@ static void rc_msg_send(struct fd_msg *msg, size_t msg_len, void *user_data)
+	}
+}
+
+static void rc_msg_postpone_rc_msg(struct fd_msg *msg, size_t msg_len,
```
`rc_msg_postpone` would be enough, or maybe just `rc_msg_push`

michallowasrzechonek-silvair commented on mesh/util.c:
```diff
@@ -50,6 +50,14 @@ void print_packet(const char *label, const void *data, uint16_t size)
 				(uint32_t) pkt_time.tv_usec/1000, label);
 }
 
+uint64_t get_timestamp_ms(void)
```
See `l_time_*` and friends.

przemyslawfierek-silvair replied:
`l_time_now` returns time since device boot ;/

michallowasrzechonek-silvair commented on mesh/model.c:
```diff
@@ -802,15 +802,6 @@ static inline void send_fd(void *data, size_t len, void *user_data)
 	fd_msg_send(user_data, data, len);
 }
 
-static inline uint64_t get_timestamp(void)
```
Separate commit

przemyslawfierek-silvair committed:
util: add function for getting current timestamp: 'get_timestamp_ms'

przemyslawfierek-silvair committed:
SP-8416: add priority queue for remote control messages

This patch adds priority queue for the Remote Control feature
to allow for better controlling time, when remote control
message should be send to the  mesh network.

michallowasrzechonek-silvair commented:
Do we have tests?

przemyslawfierek-silvair commented:
Not yet
